### DISCUSSION TITLE
[Guarantees] attribute as a way to avoid having to use the null suppressing operator

### DISCUSSION THREAD

**Discussion by kadamgreene** [dotnet/csharplang#6292] [Language Ideas] [answered]
With the introduction nullable reference types, the null suppressing operator was introduced (!) for cases where the compiler could not know that a property was not null and the developer could "guarantee" the non-null-ness of a property.  Example:

![image](https://user-images.githubusercontent.com/1278263/179834904-cf21fa83-6771-4135-83d5-c3ebe41765be.png)

If we had a way to mark Task.IsFaulted as a guarantee for Task.Exception, then we could avoid having to specify the ! operator.  Example:
```c#
[Guarantees(nameof(Exception))]
public bool IsFaulted { get; set; }
```
What made me think about this is how Typescript handles nullable types, like so:
```js
if (nullable) {
    // Typescript now knows that nullable is in fact not null
}
```
I have many times wished that I could do this in C#, but I realize that changing how `if` statements work in C# would be a very bad idea :-) So this seemed like an acceptable compromise.

I've have never written a compiler so I have no idea if this is possible / reasonable / performant / etc.

**Comment by 333fred** [comment, marked as answer]
What you're looking for is `MemberNotNullWhen`. You can put `[MemberNotNullWhen(true, nameof(Exception))]` on your `IsFaulted` property.
